engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -86,80 +94,125 @@ class EngineJobExecutionActor(replyTo: ActorRef,
-      runJob(jobDescriptor, bjeaProps)
-    case Event(CacheHit(cacheResultId), JobDescriptorData(jobDescriptor, _)) =>
-      writeToMetadata(Map(callCachingReadResultMetadataKey -> s"Cache Hit (from result ID $cacheResultId)"))
-      log.info(s"Cache hit for job ${jobDescriptor.key.call.fullyQualifiedName}, index ${jobDescriptor.key.index}! Copying cache result $cacheResultId")
-      lookupCachedResult(jobDescriptor, cacheResultId)
+      log.debug("Cache miss for job {}", jobTag)
+      runJob(data)
+    case Event(hit: CacheHit, data: ResponsePendingData) =>
+      writeToMetadata(Map(callCachingReadResultMetadataKey -> s"Cache Hit (from result ID ${hit.cacheResultId})"))
+      log.debug("Cache hit for {}! Fetching cached result {}", jobTag, hit.cacheResultId)
+      fetchCachedResults(data, jobKey.call.task.outputs, hit)
+    case Event(HashError(t), data: ResponsePendingData) =>
+      writeToMetadata(Map(callCachingReadResultMetadataKey -> s"Hashing Error: ${t.getMessage}"))
+      disableCallCaching(t)
+      runJob(data)
+  }
+
+  when(FetchingCachedOutputsFromDatabase) {
+    case Event(CachedOutputLookupSucceeded(cachedJobOutputs, cacheHit), data: ResponsePendingData) =>
+      makeBackendCopyCacheHit(cacheHit, cachedJobOutputs, data)
+    case Event(CachedOutputLookupFailed(metaInfoId, error), data: ResponsePendingData) =>
+      log.warning("Can't make a copy of the cached job outputs for {} due to {}. Running job.", jobTag, error)
+      runJob(data)
+    case Event(hashes: CallCacheHashes, data: ResponsePendingData) =>
+      addHashesAndStay(data, hashes)
+    case Event(HashError(t), data: ResponsePendingData) =>
+      disableCacheWrite(t)
+      // Can't write hashes for this job, but continue to wait for the lookup response.
+      stay
+  }
+
+  when(BackendIsCopyingCachedOutputs) {
+    // It's possible that cache writing was turned off due to a `HashError`, hence the `writeToCache` guard.
+    case Event(response: SucceededResponse, data @ ResponsePendingData(_, _, Some(Success(hashes)))) if effectiveCallCachingMode.writeToCache =>
+      saveCacheResults(hashes, response.updatedData(data))
+    case Event(response: SucceededResponse, data @ ResponsePendingData(_, _, None)) if effectiveCallCachingMode.writeToCache =>
+      // Wait for the CallCacheHashes
+      stay using response.updatedData(data)
+    case Event(response: SucceededResponse, data: ResponsePendingData) => // bad hashes or cache write off
+      saveJobCompletionToJobStore(response.updatedData(data))
+    case Event(response: BackendJobExecutionResponse, data: ResponsePendingData) =>
+      // This matches all response types other than `SucceededResponse`.
+      log.error("{}: Failed copying cache results, falling back to running job.", jobKey)
+      runJob(data)
+    case Event(hashes: CallCacheHashes, data: ResponsePendingData) =>
+      addHashesAndStay(data, hashes)
```

Comment:
This should be two events, because if data already contains a SucceededResponse we'd want to ungate ourselves and move on:
```
Event(hashes: CallCacheHashes, data @ SucceededResponseData(response, None)) => saveCacheResults(hashes, data.copy(hashes = hashes))
```
and
```
Event(hashes: CallCacheHashes, data: ResponsePendingData) => addHashesAndStay(data, hashes)
```